src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java

```diff
@@ -137,12 +158,12 @@ public CompletionStage<T> getBody() {
 
     @Override
     public void onSubscribe(Flow.Subscription subscription) {
-        this.subscription = subscription;
         // race condition with propagateError: we need to wait until
         // subscription is finished before calling onError;
         synchronized (this) {
             if (subscribed.compareAndSet(false, true)) {
-                userSubscriber.onSubscribe(subscription);
+                SubscriptionWrapper wrapped = new SubscriptionWrapper(subscription);
+                userSubscriber.onSubscribe(this.subscription = wrapped);
```

Comment:
To make it simpler to read, perhaps change it to:
```
this.subscription = new SubscriptionWrapper(subscription);
userSubscriber.onSubscribe(this.subscription);
```

It's OK with me if you prefer to have it in the current form.

Reply:
I overlooked the fact that this was a `volatile` field. So this isn't just a stylistic preference and the current form you have here adds value. So yes, looks fine to me. 
I additionally checked the generated bytecode to be sure that this does indeed do just one access to the volatile field (I wasn't sure what it would look like when you are writing to it as well as passing it as a method parameter).